File: CRGradientLabel/Objective-C/CRGradientLabel.m
//
//  CRGradientLabel.m
//  CRGradientLabel
//
//  Created by Christian Roman on 4/1/14.
//  Copyright (c) 2014 Christian Roman. All rights reserved.
//
//
//  Permission is hereby granted, free of charge, to any person obtaining a copy
//  of this software and associated documentation files (the "Software"), to deal
//  in the Software without restriction, including without limitation the rights
//  to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
//  copies of the Software, and to permit persons to whom the Software is
//  furnished to do so, subject to the following conditions:
//
//  The above copyright notice and this permission notice shall be included in
//  all copies or substantial portions of the Software.
//
//  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
//  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
//  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
//  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
//  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
//  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
//  THE SOFTWARE.
//

#import "CRGradientLabel.h"

@import CoreGraphics;

@interface CRGradientLabel ()

@end

@implementation CRGradientLabel

- (id)initWithFrame:(CGRect)frame
{
    self = [super initWithFrame:frame];
    if (self) {
        // Initialization code
    }
    return self;
}

- (void)drawRect:(CGRect)rect
{
    CGContextRef context = UIGraphicsGetCurrentContext();
    
    NSMutableArray *colors = [NSMutableArray arrayWithCapacity:[self.gradientColors count]];
    [self.gradientColors enumerateObjectsUsingBlock:^(id obj, NSUInteger idx, BOOL *stop) {
        if ([obj isKindOfClass:[UIColor class]]) {
            [colors addObject:(__bridge id)[obj CGColor]];
        } else if (CFGetTypeID((__bridge void *)obj) == CGColorGetTypeID()) {
            [colors addObject:obj];
        } else {
            @throw [NSException exceptionWithName:@"CRGradientLabelError"
                                           reason:@"Object in gradientColors array is not a UIColor or CGColorRef"
                                         userInfo:NULL];
        }
    }];
    
    CGContextSaveGState(context);
    CGContextScaleCTM(context, 1.0, -1.0);
    CGContextTranslateCTM(context, 0, -rect.size.height);
    
    CGGradientRef gradient = CGGradientCreateWithColors(NULL, (__bridge CFArrayRef)colors, NULL);
    
    CGPoint startPoint = CGPointMake(CGRectGetMidX(rect), CGRectGetMinY(rect));
    CGPoint endPoint = CGPointMake(CGRectGetMidX(rect), CGRectGetMaxY(rect));
    
    CGContextDrawLinearGradient(context, gradient, startPoint, endPoint,
                                kCGGradientDrawsAfterEndLocation | kCGGradientDrawsBeforeStartLocation);
    
    CGGradientRelease(gradient);
    CGContextRestoreGState(context);
    
    [super drawRect: rect];
}

@end

File: CRGradientLabel/Swift/CRGradientLabel.swift
//
//  CRGradientLabel.swift
//  CRGradientLabelDemo
//
//  Created by Christian Roman on 6/13/14.
//  Copyright (c) 2014 Christian Roman. All rights reserved.
//

import UIKit

class CRGradientLabel: UILabel {
    
    var gradientColors: NSArray = NSArray()
    
    override func drawRect(rect: CGRect)
    {
        var context: CGContextRef = UIGraphicsGetCurrentContext()
        
        let colors: NSMutableArray = NSMutableArray()
        
        self.gradientColors.enumerateObjectsUsingBlock({ object, index, stop in
            if object.isKindOfClass(UIColor) {
                colors.addObject(object.CGColor)
            } else if CFGetTypeID(object) == CGColorGetTypeID() {
                colors.addObject(object)
            } else {
                NSException(name: "CRGradientLabelError", reason: "Object in gradientColors array is not a UIColor or CGColorRef", userInfo: nil).raise()
            }
        })
        
        CGContextSaveGState(context)
        CGContextScaleCTM(context, 1.0, -1.0)
        CGContextTranslateCTM(context, 0, -rect.size.height)
        
        var gradient: CGGradientRef = CGGradientCreateWithColors(CGColorSpaceCreateDeviceRGB(), colors, nil)
        var startPoint: CGPoint = CGPointMake(CGRectGetMidX(rect), CGRectGetMinY(rect))
        var endPoint: CGPoint = CGPointMake(CGRectGetMidX(rect), CGRectGetMaxY(rect))
        
        CGContextDrawLinearGradient(context, gradient, startPoint, endPoint, CGGradientDrawingOptions(kCGGradientDrawsAfterEndLocation | kCGGradientDrawsBeforeStartLocation))
        
        CGContextRestoreGState(context)
        
        super.drawRect(rect)
        
    }

}
